post: self-obs (#38)
scopedb/scopedb-site: blog/2025-04-18-self-observability/index.mdx

---
slug: self-observability
title: "Self-observability of ScopeDB"
tags: [engineering]
---

> Observability is the ability to measure the internal states of a system by examining its outputs. The three pillars of observability (metrics, logs, and traces) can be used to get insights into system behavior, identify and resolve performance issues to improve system efficiency, and provide responsiveness into complex systems on bottlenecks and failures.

Author: Seems to be random changes. Keep the original words.

![Logs Details Example](logs-details-demo.png)

With detailed error logs, we can quickly locate the root cause of problems and fix them.

## Metrics in ScopeDB

Metrics are measurements of a system captured at runtime. Metrics are typically numeric values that are collected over time. They can be used to track the performance of a system, identify trends, and set alerts for specific conditions.

Generally, metrics are collected and displayed in a time-series dashboard:

![Metrics Example](metrics-demo.png)

ScopeDB reports a wide range of metrics, including counters of slow queries, ingestion and its size, gauges of memory usage, pending statements and jobs, histograms of query latencies, and so on. These metrics help us monitor ScopeDB's performance. For example, when the number of slow queries exceeds a certain threshold, an alert will be triggered to notify us.

Even more automatically, we observe the backlog of ingestion requests, scale out ScopeDB when the backlog keeps growing, and scale in when the backlog is cleared. Below is a living example when we auto-scale ScopeDB based on the ingestion backlog metrics:

> ![Metrics Workload Auto-scale Example](metrics-workload-demo.png)
> 

Author: ```suggestion
![Metrics Workload Auto-scale Example](metrics-workload-demo.png)

When the ingestion backlog grows, we temporarily instantiate additional nodes of the "scopedb-ingest" nodegroup to consume the backlog. And once the backlog gets consumed, those additional nodes are shutting down.

```

## Correlating Metrics, Traces, and Logs

To observe a complex distributed system like ScopeDB, we need to correlate metrics, traces, and logs. By correlating these three pillars of observability, we can get a complete picture of what is happening in ScopeDB.

> For example, when we see a spike in the number of slow queries in the metrics, we can list slow query logs, and then look into their associated traces. By analyzing the traces, we can identify which spans are taking a long time to complete. We can then look at the attributions or related logs for those spans to find out what is causing the delay. We can also take the workload metrics during the same time into consideration, to see if there is any correlation between the workload and the performance issues.

Reviewer: ```suggestion
For example, when there is a spike in the number of slow queries in the metrics, we can list slow query logs, and then look into the associated traces. By analyzing the traces, we’ll know which spans are taking longer to complete. Then, we could look at the attributes or related logs of these spans to find out the reason of the delay. Additionally, it is possible to take the workload metrics during the same period into consideration, and see if there is any correlation between workload and performance issues.
```

> Properly correlating metrics, traces, and logs requires the system to be instrumented correctly. This is especially challenging for a distributed system, where you need to propagate trace IDs and other metadata across different components.

Reviewer: ```suggestion
However, to properly correlate metrics, traces, and logs of a system, it requires it to be instrumented correctly. This is especially challenging in a distributed system, where trace IDs and other metadata can be propagated across a variety of components.
```

> Currently, ScopeDB leverages OpenTelemetry to instrument its codebase. ScopeDB is written in Rust, and there is no good enough distributed tracing solution at the time of writing. Thus, we developed a complete solution for distributed tracing in Rust, and open-sourced it. You can read more about the solution in this blog post: [_A Modern Approach to Distributed Tracing in Rust_](https://fast.github.io/blog/fastrace-a-modern-approach-to-distributed-tracing-in-rust/).

Reviewer: ```suggestion
Currently, ScopeDB leverages OpenTelemetry to instrument its codebase. ScopeDB is written in Rust, but at the time writing, there isn’t a sufficiently mature distributed tracing solution available. Thus, we developed a complete solution for distributed tracing in Rust, and open-sourced it. Read more about the solution in this blog post: [_A Modern Approach to Distributed Tracing in Rust_](https://fast.github.io/blog/fastrace-a-modern-approach-to-distributed-tracing-in-rust/).
```

## Looking Forward: Bootstrapping Self-observability

> In the previous introductory, we shared how an observability platform uses ScopeDB to manage metrics, logs, and trace in petabytes. A straightforward question is: can we store ScopeDB's self-observability data in ScopeDB itself? The answer is yes.

Reviewer: ```suggestion
In the previous introduction, we explored how an observability platform uses ScopeDB to store and manage metrics, logs, and trace at petabyte scale. A natural question arises: can ScopeDB store its own self-observability data? 🤯  The answer? Yes—100% it can. ⚡ 
```

> The referred observability platform is using ScopeDB for all its observability data, including ScopeDB's self-observability data. To avoid endless recursive events, it's necessary to turn of INFO level logs and traces for data ingestion of ScopeDB. In this way, when self-observability data ingested, it won't generate new events and the recursion will stop.

Reviewer: ```suggestion
The observability platform referenced in this post is using ScopeDB to store all its observability data—including ScopeDB's own. To avoid endless recursive events, it's necessary to suppress INFO level logs where query logs aside and traces for data ingestion of ScopeDB. This approach helps prevent the generation of recursive observability data when ScopeDB ingests other data.
```

Reviewer: Actually, there's no such usage "referred sth", it's always "referred to".

https://www.oxfordlearnersdictionaries.com/definition/american_english/refer-to

"reference" as a verb means "to refer to sth", which is the grammatically correct word. 

https://www.oxfordlearnersdictionaries.com/definition/english/reference_2

Also, sth that is referenced, can be shortened to sth referenced.

I didn't see the incorrectness here TBH but issues in original words.

Author: Make sense. I'll change the words on "referred" vs. "reference".

> incorrectness

"it's necessary to suppress INFO level logs where query logs aside and traces for data ingestion of ScopeDB" - this sentence subtly changes the meaning and I'd just keep the original sentence.

Author: And "The observability platform" isn't referenced "in tihs post". I'd just use a simple form "The observability platform".

Author: Done in 9d6f025.

> We're working on a demo on adapting ScopeDB into the OpenTelemetry ecosystem so that you can easily collect telemetry data into ScopeDB with minimal effort, as well as bootstrapping ScopeDB's self-observability. Stay tuned!

Reviewer: ```suggestion
Building on this, we’re working on an exciting demo that integrates ScopeDB into the OpenTelemetry ecosystem—making it easy to collect telemetry data into ScopeDB with minimal effort, while also jumpstarting its self-observability. Stay tuned—there’s more to come!
```

Author: Sounds over fancy. Keep the original words.
